### PR TITLE
sql: correctly flag cols in FKs

### PR DESCRIPTION
**Pull request by dt**
Prior to this change, even when a strict prefix of an index was used for a foreign key, all other columns in the index were considered as being used in a foreign key, not just those in said prefix.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/13439)

### CODE REVIEW

File: pkg/sql/sqlbase/structured.proto [hunk 1]
```diff
@@ -74,6 +74,7 @@ message ForeignKeyReference {
   optional uint32 index = 2 [(gogoproto.nullable) = false, (gogoproto.casttype) = "IndexID"];
   optional string name = 3 [(gogoproto.nullable) = false];
   optional ConstraintValidity validity = 4 [(gogoproto.nullable) = false];
+  optional uint32 shared_prefix_len = 5 [(gogoproto.nullable) = false];
```
Comment (vivekmenezes) [inline pkg/sql/sqlbase/structured.proto]: s/uint32/int32

File: pkg/sql/create.go [hunk 2]
```diff
@@ -1299,7 +1304,11 @@ func MakeTableDesc(
 	colsInFKs := make(map[sqlbase.ColumnID]struct{})
 	for _, idx := range desc.Indexes {
 		if idx.ForeignKey.IsSet() {
-			for i := range idx.ColumnIDs {
+			cols := len(idx.ColumnIDs)
```
Comment (vivekmenezes) [inline pkg/sql/create.go]: s/cols/numCols

File: pkg/sql/sqlbase/structured.proto (hunk 1 above)
Comment (vivekmenezes) [inline pkg/sql/sqlbase/structured.proto]: This new field probably warrants a comment.

File: pkg/sql/testdata/fk [hunk 3]
```diff
@@ -510,3 +510,25 @@ DROP TABLE otherdb.othertable
...
+
+statement ok
+CREATE TABLE domain_modules (
+  id         BIGSERIAL    NOT NULL PRIMARY KEY,
+  domain_id  BIGINT       NOT NULL,
+  module_id  BIGINT       NOT NULL,
+  CONSTRAINT domain_modules_domain_id_fk FOREIGN KEY (domain_id) REFERENCES domains (id),
+  CONSTRAINT domain_modules_module_id_fk FOREIGN KEY (module_id) REFERENCES modules (id),
+  CONSTRAINT domain_modules_uq UNIQUE (domain_id, module_id)
+);
```
Comment (vivekmenezes) [inline pkg/sql/testdata/fk]: Do these tests really test the prefix key case. I thought the FK had to reference only a subset of the primary key?

**Comment by dt** [comment]
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/create.go, line 1307 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13439#-KcKeCHq-sfy-v90j1v1-r1-1307:-KcOOZcRcHznbkpHA1xU:baio5wx) ([raw file](https://github.com/cockroachdb/cockroach/blob/59723002a940b5f8ce58f85f767f094c3e121be2/pkg/sql/create.go#L1307)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

s/cols/numCols
</blockquote></details>

 Done.

---

*[pkg/sql/sqlbase/structured.proto, line 77 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13439#-KcKeCHq-sfy-v90j1v3-r1-77:-KcOOiClZYMERtAYgg3h:b-jsplsl) ([raw file](https://github.com/cockroachdb/cockroach/blob/59723002a940b5f8ce58f85f767f094c3e121be2/pkg/sql/sqlbase/structured.proto#L77)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

This new field probably warrants a comment. 
</blockquote></details>

Done (though I really hope `len` doesn't return a negative value).

---

*[pkg/sql/testdata/fk, line 534 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13439#-KcKeCHq-sfy-v90j1v4-r1-534:-KcOPSDVNLTefQRt7gAV:b6z78y7) ([raw file](https://github.com/cockroachdb/cockroach/blob/59723002a940b5f8ce58f85f767f094c3e121be2/pkg/sql/testdata/fk#L534)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Do these tests really test the prefix key case. I thought the FK had to reference only a subset of the primary key?
</blockquote></details>

Yes, it uses a prefix of the unique index -- added a comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13439)*

**Comment by dt** [comment]
@vivekmenezes Added comments and RFAL, thanks!

**Review by jordanlewis — APPROVED** [review]

**Comment by dt** [comment]
@vivekmenezes did you get a chance to look at the comments added?

**Pull request merged**